proc/groupby.go

```diff
@@ -175,60 +179,82 @@ func (g *GroupBy) Pull() (zson.Batch, error) {
 	}
 }
 
-func (g *GroupByAggregator) createRow(ts nano.Ts, kvals [][]byte) *GroupByRow {
+func typeKey(rowkeys []zeek.TypedEncoding) string {
+	var b strings.Builder
+	for _, rowkey := range rowkeys {
+		b.WriteString(rowkey.Type.String())
+		b.WriteString(";")
+	}
+	return b.String()
+}
+
+func (g *GroupByAggregator) createRow(ts nano.Ts, rowkeys []zeek.TypedEncoding) *GroupByRow {
 	// Make a deep copy so the caller can reuse the underlying arrays.
-	kvals = append(make([][]byte, 0, len(kvals)), kvals...)
-	for k, v := range kvals {
-		if v != nil {
-			kvals[k] = append(make([]byte, 0, len(v)), v...)
-		}
+	v := make([]zeek.TypedEncoding, len(rowkeys))
+	copy(v, rowkeys)
+	key := typeKey(v)
+	_, ok := g.typeCols[key]
+	if !ok {
+		g.typeCols[key] = v
 	}
 	return &GroupByRow{
-		keyVals: kvals,
+		rowKeys: v,
 		ts:      ts,
 		columns: compile.Row{Defs: g.reducerDefs},
 	}
 }
 
+// XXX this could be made more efficient by using exporting zval.Encoding.build
+// and using it rowkey.Body.Build() and do something smarter with the zeek type
+// strings... otherwise we are sending lots of strings to the GC on each record
+// defeating the purpose of g.cacheKey.
+func (g *GroupByAggregator) key(key []byte, rowkeys []zeek.TypedEncoding) ([]byte, error) {
+	if len(rowkeys) > 0 {
+		key = append(key, rowkeys[0].Type.String()...)
+		for _, rowkey := range rowkeys[1:] {
+			key = append(key, ':')
+			key = append(key, rowkey.Type.String()...)
+		}
+		for _, rowkey := range rowkeys {
+			key = append(key, ':')
+			key = append(key, rowkey.String()...)
+		}
+	}
+	return key, nil
+}
+
 // Consume takes a record and adds it to the aggregation. Records
 // successively passed to Consume are expected to have timestamps in
 // monotonically increasing or decreasing order determined by g.reverse.
```

Comment:
This comment is out of date / should be out of date right?

Reply:
For group-by every, the comment is valid.  Until we work out timestamp sorting in zq, the plan is to just leave this broken in place.